src/phoenix/trace/openai/instrumentor.py

```diff
@@ -0,0 +1,258 @@
+    LLM_TOKEN_COUNT_TOTAL,
+    MESSAGE_CONTENT,
+    MESSAGE_FUNCTION_CALL_ARGUMENTS_JSON,
+    MESSAGE_FUNCTION_CALL_NAME,
+    MESSAGE_NAME,
+    MESSAGE_ROLE,
+    OUTPUT_MIME_TYPE,
+    OUTPUT_VALUE,
+    MimeType,
+)
+from phoenix.trace.utils import get_stacktrace, import_package
+
+from ..tracer import Tracer
+
+if TYPE_CHECKING:
+    from openai.openai_response import OpenAIResponse
+
+
+Parameters = Mapping[str, Any]
+Message = Dict[str, str]
+
+INSTRUMENTED_ATTRIBUTE_NAME = "is_instrumented_with_openinference_tracer"
+
+
+class RequestType(Enum):
+    CHAT_COMPLETION = "chat_completion"
+    COMPLETION = "completion"
+    EMBEDDING = "embedding"
+
+
+class OpenAIInstrumentor:
+    def __init__(self, tracer: Optional[Tracer] = None) -> None:
+        """Instruments your OpenAI client to automatically create spans for each API call.
+
+        Args:
+            tracer (Optional[Tracer], optional): A tracer to record and handle spans. If not
+            provided, the default tracer will be used.
+        """
+        self._tracer = tracer or Tracer()
+
+    def instrument(self) -> None:
+        """
+        Instruments your OpenAI client.
+        """
+        openai = import_package("openai")
+        is_instrumented = hasattr(
+            openai.api_requestor.APIRequestor.request,
+            INSTRUMENTED_ATTRIBUTE_NAME,
+        )
+        if not is_instrumented:
+            openai.api_requestor.APIRequestor.request = _wrap_openai_api_requestor(
+                openai.api_requestor.APIRequestor.request, self._tracer
+            )
+            setattr(
+                openai.api_requestor.APIRequestor.request,
+                INSTRUMENTED_ATTRIBUTE_NAME,
+                True,
+            )
+
+
+def _wrap_openai_api_requestor(
+    request_fn: Callable[..., Any], tracer: Tracer
+) -> Callable[..., Any]:
+    """Wraps the OpenAI APIRequestor.request method to create spans for each API call.
+
+    Args:
+        request_fn (Callable[..., Any]): The request method on openai.api_requestor.APIRequestor.
+        tracer (Tracer): The tracer to use to create spans.
+
+    Returns:
+        Callable[..., Any]: The wrapped request method.
+    """
+
+    def wrapped(*args: Any, **kwargs: Any) -> Any:
+        call_signature = signature(request_fn)
+        bound_arguments = call_signature.bind(*args, **kwargs)
+        parameters = bound_arguments.arguments["params"]
+        is_streaming = parameters.get("stream", False)
+        url = bound_arguments.arguments["url"]
+        current_status_code = SpanStatusCode.UNSET
+        events: List[SpanEvent] = []
+        attributes: SpanAttributes = dict()
+        if not is_streaming and _get_request_type(url) is RequestType.CHAT_COMPLETION:
+            for (
+                attribute_name,
+                get_parameter_attribute_fn,
+            ) in _PARAMETER_ATTRIBUTE_FUNCTIONS.items():
+                if attribute_value := get_parameter_attribute_fn(parameters):
```

Comment:
This needs to let in the number `0`. One nit-picky downside of this centralized approach is that we can't easily tailor what we want or don't want to let in when it comes to other falsy things such as `[]`, `""`, `{}`.
```suggestion
                if (attribute_value := get_parameter_attribute_fn(parameters)) is not None:
```

Reply:
Good catch.

Reply:
addressed in 27583766